Refcnt:0 error - ConnectionSetupPayload.retain() on client acceptor is ignored

```
    @Test
    void clientSetupRefCnt() {
        CloseableChannel server = RSocketServer.create().acceptor((setup, sendingSocket) -> Mono.just(new RSocket() {
        }))
                .bind(TcpServerTransport.create(8088))
                .block();

        MonoProcessor<ConnectionSetupPayload> retainedSetupPayload = MonoProcessor.create();

        RSocketConnector.create()
                .setupPayload(ByteBufPayload.create("data"))
                .acceptor((setup, sendingSocket) -> {
                    retainedSetupPayload.onNext(setup.retain());
                    return Mono.just(new RSocket() {
                    });
                })
                .connect(TcpClientTransport.create(server.address()))
                .block();

        try {
            retainedSetupPayload.as(StepVerifier::create)
                    .expectNextMatches(setup -> "data".equals(setup.getDataUtf8()))
                    .expectComplete()
                    .verify(Duration.ofSeconds(5));
        } finally {
            server.dispose();
            server.onClose().block();
        }
    }
```

Results in 
```
java.lang.AssertionError: expectation failed (failed running expectation on signal [onNext(io.rsocket.core.DefaultConnectionSetupPayload@6f9b29cf)] with [io.netty.util.IllegalReferenceCountException]:
refCnt: 0)
```

--- a/rsocket-core/src/main/java/io/rsocket/core/RSocketConnector.java
+++ b/rsocket-core/src/main/java/io/rsocket/core/RSocketConnector.java
@@ -627,9 +627,10 @@ public class RSocketConnector {
                             mtu);
 
                     return wrappedConnection
-                        .sendOne(setupFrame)
+                        .sendOne(setupFrame.retain())
                         .thenReturn(wrappedRSocketRequester);
-                  });
+                  })
+              .doFinally(signalType -> setup.release());
         });
   }
 }

--- a/rsocket-core/src/test/java/io/rsocket/core/RSocketConnectorTest.java
+++ b/rsocket-core/src/test/java/io/rsocket/core/RSocketConnectorTest.java
@@ -1,6 +1,9 @@
 package io.rsocket.core;
 
+import io.netty.buffer.ByteBuf;
+import io.netty.util.CharsetUtil;
 import io.netty.util.ReferenceCounted;
+import io.rsocket.ConnectionSetupPayload;
 import io.rsocket.Payload;
 import io.rsocket.RSocket;
 import io.rsocket.test.util.TestClientTransport;
@@ -9,9 +12,59 @@ import java.time.Duration;
 import org.assertj.core.api.Assertions;
 import org.junit.jupiter.api.Test;
 import reactor.core.publisher.Mono;
+import reactor.core.publisher.MonoProcessor;
 import reactor.test.StepVerifier;
 
 public class RSocketConnectorTest {
+
+  @Test
+  public void ensuresThatSetupPayloadCanBeRetained() {
+    MonoProcessor<ConnectionSetupPayload> retainedSetupPayload = MonoProcessor.create();
+    TestClientTransport transport = new TestClientTransport();
+
+    ByteBuf data = transport.alloc().buffer();
+
+    data.writeCharSequence("data", CharsetUtil.UTF_8);
+
+    RSocketConnector.create()
+        .setupPayload(ByteBufPayload.create(data))
+        .acceptor(
+            (setup, sendingSocket) -> {
+              retainedSetupPayload.onNext(setup.retain());
+              return Mono.just(new RSocket() {});
+            })
+        .connect(transport)
+        .block();
+
+    Assertions.assertThat(transport.testConnection().getSent())
+        .hasSize(1)
+        .first()
+        .matches(
+            bb -> {
+              DefaultConnectionSetupPayload payload = new DefaultConnectionSetupPayload(bb);
+              return !payload.hasMetadata() && payload.getDataUtf8().equals("data");
+            })
+        .matches(buf -> buf.refCnt() == 2)
+        .matches(
+            buf -> {
+              buf.release();
+              return buf.refCnt() == 1;
+            });
+
+    retainedSetupPayload
+        .as(StepVerifier::create)
+        .expectNextMatches(
+            setup -> {
+              String dataUtf8 = setup.getDataUtf8();
+              return "data".equals(dataUtf8) && setup.release();
+            })
+        .expectComplete()
+        .verify(Duration.ofSeconds(5));
+
+    Assertions.assertThat(retainedSetupPayload.peek().refCnt()).isZero();
+
+    transport.alloc().assertHasNoLeaks();
+  }
 
   @Test
   public void ensuresThatMonoFromRSocketConnectorCanBeUsedForMultipleSubscriptions() {

